From c14e1e7edb97c47ae8ded27c07f0551bc45172c4 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Wed, 26 Nov 2025 12:03:27 +0100
Subject: [PATCH] usb: dwc2: only enable wake irq when the hw is enabled for
 wakeup

The requirement for wakeup are fine-checked, in
dwc2_gadget_can_poweroff_phy() and dwc2_host_can_poweroff_phy(dwc2)).
The device_may_wakeup(dev) || device_wakeup_path(dev), may let the
driver enable the wake IRQ, even if no wakeup event is expected.
So align wake irq with the HW state, based on above helpers.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I8de0ae25c1e59e482f752509aecd1b7b28c62080
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512994
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/dwc2/platform.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/dwc2/platform.c b/drivers/usb/dwc2/platform.c
index f38c404eb3f4..8b0e06600369 100644
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -981,7 +981,8 @@ static int __maybe_unused dwc2_suspend(struct device *dev)
 		dwc2->phy_off_for_suspend = true;
 	}
 
-	if (device_may_wakeup(dev) || device_wakeup_path(dev))
+	/* If HW has been kept enabled for wakeup, enable wake irq */
+	if (dwc2->ll_hw_enabled && !dwc2->phy_off_for_suspend)
 		enable_irq_wake(dwc2->irq);
 
 	return ret;
@@ -996,7 +997,7 @@ static int __maybe_unused dwc2_resume(struct device *dev)
 	if (!dwc2->ll_hw_enabled)
 		return 0;
 
-	if (device_may_wakeup(dev) || device_wakeup_path(dev))
+	if (!dwc2->phy_off_for_suspend)
 		disable_irq_wake(dwc2->irq);
 
 	if (!ll_hw_enabled) {
-- 
2.53.0

